aten/src/ATen/native/native_functions.yaml

```diff
@@ -3847,6 +3847,27 @@
   dispatch:
     CompositeExplicitAutograd: slice
 
+- func: slice_scatter(Tensor self, Tensor mutated_view, int dim=0, int? start=None, int? end=None, int step=1) -> Tensor
+  variants: function, method
+  device_check: NoCheck
+  device_guard: False
+  dispatch:
+    CPU, CUDA: slice_scatter
```

Comment:
Why limit these to these two devices? Why not a CompositeAutogradExplicit?

Reply:
So, I don't think we want other backends like xla/vulkan to use this implementation for two reasons:
- `slice_scatter` is used to implement the functionalization pass: when called inside of the pass, it's currently a primitive w.r.t the functionalization pass. Its implementation happens to use views and mutations, but that should be an implementation detail that doesn't leak into the pass, since the pass **uses** slice_scatter to remove aliasing in the first place.
- `slice_scatter` (for cpu/cuda at least) is implemented with `slice`, a view op. vulkan and xla (eventually) won't have actual implementations of these ops, since they'll rely on the functionalization pass to remove view ops (and replace them with view_copy calls).


But I feel like there's a mismatch between `slice_scatter` "the public API" and `slice_scatter` "in the implementation weeds of functionalization". If a user calls `slice_scatter` directly, it seems reasonable to make it composite - it should decompose into a `slice` and a `copy_` call, which will then get fed into the functionalization pass if you're running on xla. But when it's used inside of the functionalization pass, we don't want it to (and it currently doesn't) get recursively fed into the functionalization pass again (?).

Let me know if that makes sense (@ezyang). It seems to me like the options are (in order of implementation ease):
- Keep as a public API, only implemented for cpu/cuda (the code deserves a comment if we do this)
- Add two versions of the op: one thats public and composite, and one that's private and only used by the functionalization pass (this would make the logic uglier though, since we'd need to special-case these ops in the codegen)

Reply:
I think your point is valid and indeed backends like xla/vulkan won't be able to use this implementation and will have to specify their own.
But I don't think that's a reason to make all other backends (mkl/mlc/etc) have to implement it as well while this default implementation would work.

Just to be clear, I suggested a composite Explicit implementation, not implicit. So it will behave exactly the same as it does right now, just that other backends will also be able to use that by default if they support view/inplace!

Reply:
Ok, I think that's fair.

The main thing I'm worried about is error detection: a backend opting into the functionalization pass, and getting silent failures because they forgot to register their own kernels for `slice_scatter` et.al.

We can probably worry about that for when I actually try to get vulkan/xla working with the pass though. At the very least, there are only two backends that we know want to use the pass. To harden it for future backends, maybe we'll want separate private API's for these functions in the future.